plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/data/MapboxViewportDataSource.kt

```diff
@@ -0,0 +1,272 @@
+          pointsForFollowing,
+          followingPadding,
+          followingBearingProperty.get(),
+          followingPitchProperty.get()
+        )
+      } else {
+        val mapSize = mapTransform.getSize()
+        val screenBox = getScreenBoxForFraming(mapSize, followingPadding)
+        val cameraState = cameraManager.cameraState
+        val fallbackCameraOptions = CameraOptions.Builder()
+          .center(pointsForFollowing.firstOrNull() ?: cameraState.center)
+          .padding(getMapAnchoredPaddingFromUserPadding(mapSize, followingPadding))
+          .bearing(followingBearingProperty.get())
+          .pitch(followingPitchProperty.get())
+          .zoom(cameraState.zoom)
+          .build()
+        if (pointsForFollowing.size > 1) {
+          cameraManager.cameraForCoordinates(
+            pointsForFollowing,
+            fallbackCameraOptions,
+            screenBox
+          )
+        } else {
+          fallbackCameraOptions
+        }
+      }
+
+      followingCenterProperty.fallback = cameraFrame.center!!
+      options.followingFrameOptions.run {
+        followingZoomProperty.fallback = max(min(cameraFrame.zoom!!, maxZoom), minZoom)
+      }
+      appliedFollowingPadding = cameraFrame.padding!!
+      return
+    }
+    // nothing to frame
+    options.followingFrameOptions.run {
+      val cameraState = cameraManager.cameraState
+      followingBearingProperty.fallback = cameraState.bearing
+      followingPitchProperty.fallback = defaultPitch
+      followingCenterProperty.fallback = cameraState.center
+      followingZoomProperty.fallback = max(min(cameraState.zoom, maxZoom), minZoom)
+    }
+  }
+
+  private fun updateOverviewData() {
+    targetLocation?.let { nonNullTargetLocation ->
+      val pointsForOverview = listOf(nonNullTargetLocation.toPoint())
+
+      overviewBearingProperty.fallback = normalizeBearing(
+        cameraManager.cameraState.bearing,
+        BEARING_NORTH
+      )
+
+      val cameraFrame = if (pointsForOverview.isNotEmpty()) {
+        cameraManager.cameraForCoordinates(
+          pointsForOverview,
+          overviewPadding,
+          overviewBearingProperty.get(),
+          overviewPitchProperty.get()
+        )
+      } else {
+        cameraManager.cameraState.toCameraOptions()
+      }
+
+      overviewCenterProperty.fallback = cameraFrame.center!!
+      overviewZoomProperty.fallback = min(
+        cameraFrame.zoom!!,
+        options.overviewFrameOptions.maxZoom
+      )
+      return
+    }
+    // nothing to frame
+    options.overviewFrameOptions.run {
+      val cameraState = cameraManager.cameraState
+      overviewBearingProperty.fallback = cameraState.bearing
+      overviewPitchProperty.fallback = cameraState.pitch
+      overviewCenterProperty.fallback = cameraState.center
+      overviewZoomProperty.fallback = min(cameraState.zoom, maxZoom)
+    }
+  }
+
+  private fun Location.toPoint(): Point = Point.fromLngLat(longitude, latitude)
```

Comment:
If there are more things like that - move to Util class?

Reply:
good idea